blscurve/bls_batch_verifier.nim

```diff
@@ -0,0 +1,372 @@
+    ## A (Public Key, Message, Signature) triplet
+    ## that will be batch verified.
+    ## This should not hold GC-ed memory
+    ## as this would complexify multithreading.
+    ## Consequently this assumes that message
+    ## is the output of a fixed size hash function.
+    signature: Signature
+    pubkey: PublicKey
+    message: array[32, byte]
+
+  BatchedBLSVerifier* = object
+    ## A type to batch BLS multi signatures (aggregated or individual)
+    ## verification using multiple cores if compiled with OpenMP
+    sets: seq[SignatureSet]
+
+    # Per-batch contexts for multithreaded batch verification
+    batchContexts: seq[ContextMultiAggregateVerify[DST]]
+    updateResults: seq[tuple[ok: bool, padCacheLine: array[64, byte]]]
+
+func init*(T: type BatchedBLSVerifier): T {.inline.} =
+  ## Initialize or reinitialize a batchedBLS Verifier
+  discard # default initialization
+
+func clear*(batcher: var BatchedBLSVerifier) {.inline.} =
+  ## Initialize or reinitialize a batchedBLS Verifier
+  batcher.sets.setLen(0)
+  batcher.batchContexts.setLen(0)
+
+func incl*(
+       batcher: var BatchedBLSVerifier,
+       public_key: PublicKey,
+       message: array[32, byte],
+       signature: Signature
+     ): bool {.inline.} =
+  ## Include a (public key, message, signature) triplet
+  ## to the batch for verification.
+  ##
+  ## Always return true, as it assumes subgroup checks
+  ## and infinity checks were done prior.
+  ## This might change in the future via an optinal check parameter
+  ## so the return value should be checked
+  ##
+  ## The proof-of-possession MUST be verified before calling this function.
+  ## It is recommended to use the overload that accepts a proof-of-possession
+  ## to enforce correct usage.
+  batcher.sets.add SignatureSet(
+    signature: signature,
+    pubkey: public_key,
+    message: message
+  )
+  return true
+
+func incl*(
+       batcher: var BatchedBLSVerifier,
+       public_keys: openarray[PublicKey],
+       message: array[32, byte],
+       signature: Signature
+     ): bool =
+  ## Include a (array of public keys, message, signature) triplet
+  ## to the batch for verification.
+  ##
+  ## All public keys sign the same message
+  ## and signature is their aggregated signature
+  ##
+  ## Returns false if no public keys are passed
+  ## Returns true otherwise
+  ##
+  ## The proof-of-possession MUST be verified before calling this function.
+  ## It is recommended to use the overload that accepts a proof-of-possession
+  ## to enforce correct usage.
+  if publicKeys.len == 0:
+    return false
+
+  var aggAffine{.noInit.}: PublicKey
+  if not aggAffine.aggregateAll(publicKeys):
+    return false
+
+  batcher.sets.add SignatureSet(
+    signature: signature,
+    pubkey: aggAffine,
+    message: message
+  )
+
+  return true
+
+# Serial Batch Verifier
+# ----------------------------------------------------------------------
+
+func batchVerifySerial*(
+       batcher: BatchedBLSVerifier,
```

Comment:
why not simply take an `openArray[SignatureSet]` here?

Reply:
This is similar to why we have `init`/`update`/`finish` for sha256.

This gives us the flexibility to collect signature sets from multiple procs and we can always create a `batchVerify` primitive on top of `init`/`incl`/`batchVerify` primitives.

Reply:
but I can do the same with a `seq`? ie there's nothing in batchedblsverifier that needs init/update/finish really (unlike hashing that has an active init that initializes the hash with a seed).. 

also, if memory reallocation is a thing, shouldn't the serial verifier reuse the `ContextMultiAggregateVerify`  from the passed-in instace? though again, I'd separate that into a separate cache to avoid accidentally accumuilating more and more signatures

Reply:
When I designed it, I wanted so that during state processing you pass the BatchVerify object and replace `verify` in attester_slashings, attestations, proposer_slashings by `batcher.incl(...)`. It's basically a seq inside.

We can't use the `ContextMultiAggregateVerify` for 2 reasons, for multithreading we need one per core, and second it's 3+MB stack object.

Reply:
what I mean is:
```
type BlsCache = object
  cmav: seq[CMAV]

batchVerifySerial(sigs: openArray[sigset], cache: BlsCache) =
  cache.cmav.setLen(1)
  cache.cmav[0].docontextthing()
```

this has several advantages: 
* there's less risk of forgetting to clear the old sigset
* it delineates the cache (long-lived) from the set of signatures to check (short-lived) in separate types
* there's no weird incl that works like `add` but is named like a set operation (it doesn't guarantee uniqueness, so why is it called `incl`?)
* it's clear that the cache is a cache that can either be reused or not depending on desired tradeoffs
* same cache type can be used for both multi and single threaded - I imagine that we might want to provide a single-threaded command line option that doesn't use the extra threads - this mode is important when sharing resources with other processes - the multicpu mode will always carry a significant overhead.